runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java

```diff
@@ -191,17 +272,41 @@ public Object applyUpdateUnsafe(SMREntry entry, boolean isOptimistic) {
                     undoRecordFunctionMap
                             .get(entry.getSMRMethod());
             if (undoRecordTarget != null) {
-                entry.setUndoRecord(undoRecordTarget
-                        .getUndoRecord(object, entry.getSMRArguments()));
-                entry.setUndoable(true);
-                if (isOptimistic && optimisticallyUndoable) {
+                // calculate the undo record if it doesn't exist.
+                if (!entry.isUndoable()) {
+                    entry.setUndoRecord(undoRecordTarget
+                            .getUndoRecord(object, entry.getSMRArguments()));
+                    entry.setUndoable(true);
+                }
+                // If this is a standard mutation record
+                // and (1) we are not in optimistic mode
+                // (2) the undoLog is not empty OR
+                // the undoLog is empty and we need to
+                // generate undo records add an undo
+                // record to the log.
+                if (!isOptimistic && (!undoLog.isEmpty() ||
+                        needUndoRecordUnsafe())) {
+                    undoLog.addFirst(entry);
+                }
+                // If we're in optimistic mode, add us to the
+                // optimistic undo log. (If there's a point. If
+                // the object isn't optimistically undoable anymore
+                // there's no point.)
+                else if (isOptimistic && optimisticallyUndoable) {
                     optimisticUndoLog.addFirst(entry);
                 }
             } else {
+                // We can't generate an undo record, so clear the
+                // undo logs, and mark that the object is
+                // no longer optimistically undoable if we
+                // are in optimistic mode.
+                undoLog.clear();
```

Comment:
this seems too restrictive. If we are in optimistic mode, shouldn't we clear the optimisticUndoLog only (if exists) and not touch the undoLog?